File: 1207-delete-nodes-and-return-forest/delete-nodes-and-return-forest.cpp
/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode() : val(0), left(nullptr), right(nullptr) {}
 *     TreeNode(int x) : val(x), left(nullptr), right(nullptr) {}
 *     TreeNode(int x, TreeNode *left, TreeNode *right) : val(x), left(left), right(right) {}
 * };
 */
class Solution {
public:
    TreeNode* dfs(TreeNode* root, unordered_set<int> &s,vector<TreeNode*> &v)
    {
        if(root==NULL)
        return root;

        root->left=dfs(root->left,s,v);
        root->right=dfs(root->right,s,v);

        if(s.find(root->val)!=s.end())
        {
            if(root->left!=NULL) v.push_back(root->left);
            if(root->right!=NULL) v.push_back(root->right);
            delete root;
            root=NULL; 
        }
        return root;
    }

    vector<TreeNode*> delNodes(TreeNode* root, vector<int>& to_delete) 
    {
        unordered_set<int> s(to_delete.begin(),to_delete.end());
        vector<TreeNode*> v;

        if(s.find(root->val)==s.end())  v.push_back(root);

        dfs(root,s,v);
        return v;
    }
};

File: 1207-delete-nodes-and-return-forest/delete-nodes-and-return-forest.java
/**
 * Definition for a binary tree node.
 * public class TreeNode {
 *     int val;
 *     TreeNode left;
 *     TreeNode right;
 *     TreeNode() {}
 *     TreeNode(int val) { this.val = val; }
 *     TreeNode(int val, TreeNode left, TreeNode right) {
 *         this.val = val;
 *         this.left = left;
 *         this.right = right;
 *     }
 * }
 */
class Solution {
    private TreeNode solve(TreeNode root, Set<Integer> deleteSet, List<TreeNode> res){
        if(root == null) return root;

        root.left = solve(root.left, deleteSet, res);
        root.right = solve(root.right, deleteSet, res);

        if(deleteSet.contains(root.val)){
            if(root.left != null){
                res.add(root.left);
            }
            if(root.right != null){
                res.add(root.right);
            }
            return null;
        }

        return root;
    }

    public List<TreeNode> delNodes(TreeNode root, int[] to_delete) {
        Set<Integer> deleteSet = new HashSet<>();
        for (int val : to_delete) {
            deleteSet.add(val);
        }

        List<TreeNode> res = new ArrayList<>();
        solve(root, deleteSet, res);
        if(!deleteSet.contains(root.val)){
            res.add(root);
        }
        return res;
    }
}